source/crud/bulk-write.md

```diff
@@ -0,0 +1,778 @@
+documents have the following format:
+
+#### Insert
+
+```json
+{
+    "insert": <Int32>,
+    "document": <Document>
+}
+```
+
+If the document to be inserted does not contain an `_id` field, drivers MUST generate a new
+[`ObjectId`](../objectid.rst) and add it as the `_id` field at the beginning of the document.
+
+#### Update
+
+```json
+{
+    "update": <Int32>,
+    "filter": <Document>,
+    "updateMods": <Document | Array>,
+    "multi": Optional<Boolean>,
+    "upsert": Optional<Boolean>,
+    "arrayFilters": Optional<Array>,
+    "hint": Optional<Document | String>
+}
+```
+
+#### Delete
+
+```json
+{
+    "delete": <Int32>,
+    "filter": <Document>,
+    "multi": Optional<Boolean>,
+    "hint": Optional<Document | String>,
+    "collation": Optional<Document>
+}
+```
+
+### Namespace Information
+
+The `nsInfo` field is an array containing the namespaces on which the write operations should be
+performed. Drivers MUST NOT include duplicate namespaces in this list. The documents in the
+`nsInfo` array have the following format:
+
+```json
+{
+    "ns": <String>
+}
+```
+
+### `errorsOnly` and `verboseResults`
+
+The `errorsOnly` field indicates whether the results cursor returned in the `bulkWrite` response
+should contain only errors and omit individual results. If false, both individual results for
+successful operations and errors will be returned. This field is optional and defaults to false on
+the server.
+
+`errorsOnly` corresponds to the `verboseResults` option defined on `BulkWriteOptions`. If the user
+specified a value for `verboseResults`, drivers MUST define `errorsOnly` as the opposite of
+`verboseResults`. If the user did not specify a value for `verboseResults`, drivers MUST define
+`errorsOnly` as `true`.
+
+### `ordered`
+
+The `ordered` field defines whether writes should be executed in the order in which they were
+specified, and, if an error occurs, whether the server should halt execution of further writes. It
+is optional and defaults to true on the server. Drivers MUST explicitly define `ordered` as `true`
+in the `bulkWrite` command if a value is not specified in `BulkWriteOptions`. This is required to
+avoid inconsistencies between server and driver behavior if the server default changes in the
+future.
+
+### Size Limits
+
+The server reports a `maxBsonObjectSize` in its `hello` response. This value defines the maximum
+size for documents that are inserted into the database. Documents that are sent to the server but
+are not intended to be inserted into the database (e.g. command documents) have a size limit of
+`maxBsonObjectSize + 16KiB`. When an acknowledged write concern is used, drivers MUST NOT perform
+any checks related to these size limits and MUST rely on the server to raise an error if a limit is
+exceeded. However, when an unacknowledged write concern is used, drivers MUST raise an error if one
+of the following limits is exceeded:
```

Comment:
Arguments I have for not checking these sizes for acknowledged writes:

- Relying on the server to return an error for oversized documents aligns with [this](https://github.com/mongodb/specifications/?tab=readme-ov-file#where-possible-depend-on-server-to-return-errors) driver specification mantra.
- Not checking sizes driver-side means the server can still proceed with other valid writes in the bulk write.
- If we have a hard cap of `maxMessageSizeBytes` for the total message size, acknowledged or unacknowledged, users will still be limited to what the server deems an acceptable number of bytes to be sent in one message.

Would additional language here to clarify that drivers should never send a message larger than `maxMessageSizeBytes` address your concerns about potential malicious input to this method? I ask because it seems like users do value their other writes being performed when an oversized document is specified, e.g. [this recent forum post](https://www.mongodb.com/community/forums/t/pymongo-insert-many-raising-operationfailure-instead-of-bulkwriteerror-on-oversized-documents/270467).

Reply:
> Not checking sizes driver-side means the server can still proceed with other valid writes in the bulk write.

Right, I forgot that point.

> Would additional language here to clarify

No need IMO. I think the batch splitting already addresses that messages will not exceed `maxMessageSizeBytes`.